lib/hterm.js

```diff
@@ -212,6 +151,102 @@ hterm.ScrollPort.prototype.setBackgroundColor = function () {
+  caret.addEventListener('compositionupdate', () => {
+    this.cursorNode_.style.backgroundColor = 'yellow';
+    this.cursorNode_.style.borderColor = 'yellow';
+  });
+
+  // at this point the char(s) is ready
+  caret.addEventListener('compositionend', () => {
+    ongoingComposition = false;
+    this.cursorNode_.style.backgroundColor = '';
+    this.setCursorShape(this.getCursorShape());
+    this.cursorNode_.style.borderColor = this.getCursorColor();
+    caret.innerText = '';
+  });
+
+  // if you open the `Emoji & Symbols` (ctrl+cmd+space)
+  // and select an emoji, it'll be inserted into our caret
+  // and stay there until you star a compositon event.
+  // to avoid that, we'll just check if there's an ongoing
+  // compostion event. if there's one, we do nothing.
+  // otherwise, we just remove the emoji and stop the event
+  // propagation.
+  // PS: this event will *not* be fired when a standard char
+  // (a, b, c, 1, 2, 3, etc) is typed – only for composed
+  // ones and `Emoji & Symbols`
+  caret.addEventListener('input', e => {
+    if (!ongoingComposition) {
+      caret.innerText = '';
+      e.stopPropagation();
+      e.preventDefault();
+    }
+  });
+
+  // we need to capture pastes, prevent them and send its contents to the terminal
+  caret.addEventListener('paste', e => {
+    e.stopPropagation();
+    e.preventDefault();
+    const text = e.clipboardData.getData('text');
+    this.onVTKeystroke(text);
+  });
+
+  // here we replicate the focus/blur state of our caret on the `hterm` caret
+  caret.addEventListener('focus', () => {
+    this.cursorNode_.setAttribute('focus', true);
+    this.restyleCursor_();
+  });
+  caret.addEventListener('blur', () => {
+    this.cursorNode_.setAttribute('focus', false);
+    this.restyleCursor_();
+  });
+
+  // this is necessary because we need to access the `document_` and the hyperCaret
+  // on `hterm.Screen.prototype.syncSelectionCaret`
+  this.primaryScreen_.terminal = this;
+  this.alternateScreen_.terminal = this;
+};
+
+// ensure that our contenteditable caret is injected
+// inside the term's cursor node and that it's focused
+hterm.Terminal.prototype.focusHyperCaret = function () {
+  if (!this.hyperCaret.parentNode !== this.cursorNode_) {
+    this.cursorNode_.appendChild(this.hyperCaret);
+  }
+  this.hyperCaret.focus();
+};
+
+hterm.Screen.prototype.syncSelectionCaret = function () {
+  const p = this.terminal.hyperCaret;
+  const doc = this.terminal.document_;
+  const win = doc.defaultView;
+  const s = win.getSelection();
+  const r = doc.createRange();
+  r.selectNodeContents(p);
+  s.removeAllRanges();
+  s.addRange(r);
+};
+
+// fixes a bug in hterm, where the cursor goes back to `BLOCK`
+// after the bell rings
+const oldRingBell = hterm.Terminal.prototype.ringBell;
+hterm.Terminal.prototype.ringBell = function () {
+  oldRingBell.call(this);
+  setTimeout(() => {
+    this.restyleCursor_();
+  }, 200);
```

Comment:
@rauchg I'm not sure if it's necessary – it's just an extension of the timeout that `hterm` creates:

```javascript
hterm.Terminal.prototype.ringBell = function() {
  this.cursorNode_.style.backgroundColor =
      this.scrollPort_.getForegroundColor();

  var self = this;
  setTimeout(function() {
      self.cursorNode_.style.backgroundColor = self.prefs_.get('cursor-color');
    }, 200);
...
```